test/tests/text_highlighting_test.py

```diff
@@ -78,6 +83,90 @@
+    Morpheme(lemma="番", inflection="番", highest_inflection_learning_interval=10),
+    Morpheme(lemma="お", inflection="お", highest_inflection_learning_interval=10),
+    Morpheme(lemma="見事", inflection="見事", highest_inflection_learning_interval=30),
+    Morpheme(lemma="腕", inflection="腕", highest_inflection_learning_interval=10),
+    Morpheme(
+        lemma="変わら", inflection="変わら", highest_inflection_learning_interval=10
+    ),
+    Morpheme(lemma="だ", inflection="だ", highest_inflection_learning_interval=10),
+    Morpheme(lemma="だっ", inflection="だっ", highest_inflection_learning_interval=10),
+    Morpheme(lemma="ああ", inflection="ああ", highest_inflection_learning_interval=10),
+    Morpheme(lemma="ず", inflection="ず", highest_inflection_learning_interval=10),
+    Morpheme(lemma="た", inflection="た", highest_inflection_learning_interval=10),
+    Morpheme(lemma="相", inflection="相", highest_inflection_learning_interval=10),
+]
+
+##############################################################################################
+#                                    CASE: Morph and ruby interaction
+##############################################################################################
+# This third example sets all the cases where morphs and rubies coexist.
+# |-mmm---| |--mmm--| |---mmm-| |-mmm---| |-mmmmm-| |-mmmmm-| |--mmm--| |--m-m--|
+# |----rrr| |--rrr--| |--rrr--| |--rrr--| |--rrr--| |--r-r--| |-rrrrr-| |-rrrrr-|
+# Collection choice is arbitrary.
+# Database choice is arbitrary.
+##############################################################################################
+case_morph_and_ruby = FakeEnvironmentParams(
+    config=config_big_japanese_collection,
+)
+CASE_MORPH_AND_RUBY_INPUT_TEXT = "12345  09876[def]  12345[abc]  12[abc] 34[abc]5  09876[def] 1 23[abc]45  012345[abc]  1234512345[abc]  0123[abc]45 1 23456[abc]  12345777[zyzzzzz]"
+CASE_MORPH_AND_RUBY_CORRECT_OUTPUT = (
+    ""
+    + '<span morph-status="unknown">12345</span>'
+    + " "
+    + "<ruby>09876<rt>def</rt></ruby>"
+    + " "
+    + '<span morph-status="unknown"><ruby>12345<rt>abc</rt></ruby></span>'
+    + " "
+    + '<span morph-status="unknown"><ruby>12<rt>abc</rt></ruby><ruby>34<rt>abc</rt></ruby>5</span>'
+    + " "
+    + "<ruby>09876<rt>def</rt></ruby>"
+    + " "
+    + '<span morph-status="unknown">1<ruby>23<rt>abc</rt></ruby>45</span>'
+    + " "
+    + '<ruby>0<span morph-status="unknown">12345</span><rt>abc</rt></ruby>'
+    + " "
+    + '<ruby><span morph-status="unknown">12345</span><span morph-status="unknown">12345</span><rt>abc</rt></ruby>'
+    + " "
+    + '<ruby>0<span morph-status="unknown">123</span><rt>abc</rt></ruby><span morph-status="unknown">45</span>'
+    + " "
+    + '<span morph-status="unknown">1</span><ruby><span morph-status="unknown">2345</span>6<rt>abc</rt></ruby>'
+    + " "
+    + '<ruby><span morph-status="unknown">12345</span><span morph-status="unknown">777</span><rt>zyzzzzz</rt></ruby>'
+)
```

Comment:
I will also collapse this into one row, I had it this way for clarity while adding test cases and debugging.